### PR TITLE
feat: Implement `OPTIONAL MATCH`

### PR DESCRIPTION
**Pull request by emotionbug**

### CODE REVIEW

**Comment by emotionbug** [comment]
Second commit is for fixing bug on `ORDER BY` clause.

Thanks @JoshInnis

File: src/backend/parser/cypher_gram.y [hunk 1]
```diff
@@ -118,7 +118,7 @@
 
 /* DELETE clause */
 %type <node> delete
-%type <boolean> detach_opt
+%type <boolean> optional_opt detach_opt
```
Comment (JoshInnis) [inline src/backend/parser/cypher_gram.y:121]: The optional_opt should be defined with the rest of the match clause items starting on line 111,

File: regress/expected/cypher_match.out [hunk 2]
```diff
@@ -889,6 +889,41 @@ $$) AS (i agtype);
  {"id": 1688849860263938, "label": "v2", "properties": {"id": "middle"}}::vertex
 (3 rows)
 
+--
```
Comment (JoshInnis) [inline regress/expected/cypher_match.out:892]: There needs to be test cases for when Optional Match is the first clause in a cypher statement.

Comment (emotionbug) [reply regress/expected/cypher_match.out:892]: OK. Added more tests.

File: src/backend/parser/cypher_gram.y (hunk 1 above)
Comment (emotionbug) [reply src/backend/parser/cypher_gram.y]: Moved on MATCH clause.

File: src/backend/parser/cypher_clause.c [hunk 3]
```diff
@@ -2443,19 +2647,15 @@ static transform_entity *transform_VLE_edge_entity(cypher_parsestate *cpstate,
...
-    FuncCall *func = NULL;
     Alias *alias = NULL;
     Node *var = NULL;
     transform_entity *vle_entity = NULL;
 
     /* it better be a function call node */
     Assert(IsA(rel->varlen, FuncCall));
 
-    /* get the function */
-    func = (FuncCall*)rel->varlen;
```
Comment (JoshInnis) [inline src/backend/parser/cypher_clause.c:2455]: Needs to be removed. The updated VLE patch will take care of this

File: src/backend/parser/cypher_clause.c [hunk 4]
```diff
@@ -1320,6 +1325,161 @@ static Query *transform_cypher_match(cypher_parsestate *cpstate,
...
+    int			i;
+    ParseNamespaceItem *nsitem;
+
+    j->jointype = JOIN_LEFT;
+
+
+    l_alias = makeAlias(CYPHER_OPT_LEFT_ALIAS, NIL);
+    r_alias = makeAlias(CYPHER_OPT_RIGHT_ALIAS, NIL);
+
+    j->larg = transform_clause_for_join(cpstate, clause->prev, &l_rte,
```
Comment (JoshInnis) [inline src/backend/parser/cypher_clause.c:1434]: Optional Match may be the first clause in a query. There are no cases to handle that situation

File: src/backend/parser/cypher_clause.c (hunk 3 above)
Comment (emotionbug) [reply src/backend/parser/cypher_clause.c]: Sorry, I deleted it. When merging, inserted it due to a compilation warning.

**Comment by jrgemignani** [comment]
There are conflicts that need to be resolved since the master branch has changed.

**Comment by emotionbug** [comment]
@jrgemignani Resolved.

**Comment by emotionbug** [comment]
Commit details have become too long, so organize and open a new PR.

**Pull request closed without merging**